File: authserver/src/main/java/l2s/authserver/network/gamecomm/GameServerDescription.java
package l2s.authserver.network.gamecomm;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.CopyOnWriteArraySet;

public class GameServerDescription {
    private final Map<Integer, HostInfo> _hosts;
    private int _serverType;
    private int _ageLimit;
    private int _protocol;
    private boolean _isOnline;
    private boolean _isPvp;
    private boolean _isShowingBrackets;
    private boolean _isGmOnly;
    private int _maxPlayers;

    private final Set<String> _accounts;

    private final String ipAddress;

    public GameServerDescription(String ipAddress) {
        this.ipAddress = ipAddress;
        _hosts = new HashMap<>();
        _accounts = new CopyOnWriteArraySet<>();
    }

    public void addHost(HostInfo host) {
        _hosts.put(host.getId(), host);
    }

    public HostInfo removeHost(int id) {
        return _hosts.remove(id);
    }

    public HostInfo[] getHosts() {
        return _hosts.values().toArray(new HostInfo[0]);
    }

    public void setMaxPlayers(int maxPlayers) {
        _maxPlayers = maxPlayers;
    }

    public int getMaxPlayers() {
        return _maxPlayers;
    }

    public int getOnline() {
        return _accounts.size();
    }

    public Set<String> getAccounts() {
        return _accounts;
    }

    public void addAccount(String account) {
        _accounts.add(account);
    }

    public void removeAccount(String account) {
        _accounts.remove(account);
    }

    public int getServerType() {
        return _serverType;
    }

    public boolean isOnline() {
        return _isOnline;
    }

    public void setOnline(boolean online) {
        _isOnline = online;
    }

    public void setServerType(int serverType) {
        _serverType = serverType;
    }

    public boolean isPvp() {
        return _isPvp;
    }

    public void setPvp(boolean pvp) {
        _isPvp = pvp;
    }

    public boolean isShowingBrackets() {
        return _isShowingBrackets;
    }

    public void setShowingBrackets(boolean showingBrackets) {
        _isShowingBrackets = showingBrackets;
    }

    public boolean isGmOnly() {
        return _isGmOnly;
    }

    public void setGmOnly(boolean gmOnly) {
        _isGmOnly = gmOnly;
    }

    public int getAgeLimit() {
        return _ageLimit;
    }

    public void setAgeLimit(int ageLimit) {
        _ageLimit = ageLimit;
    }

    public int getProtocol() {
        return _protocol;
    }

    public void setProtocol(int protocol) {
        _protocol = protocol;
    }

    public String getIpAddress() {
        return ipAddress;
    }

    public void setDown() {
        setOnline(false);
        _accounts.clear();
    }

    public static class HostInfo {
        private final int _id;
        private final String _ip;
        private final String _innerIP;
        private final int _port;
        private final String _key;

        public HostInfo(int id, String ip, String innerIP, int port, String key) {
            _id = id;
            _ip = ip;
            _innerIP = innerIP;
            _port = port;
            _key = key;
        }

        public int getId() {
            return _id;
        }

        public String getIP() {
            return _ip;
        }

        public String getInnerIP() {
            return _innerIP;
        }

        public int getPort() {
            return _port;
        }

        public String getKey() {
            return _key;
        }
    }
}

File: authserver/src/main/java/l2s/authserver/network/gamecomm/GameServerDescription.kt
package l2s.authserver.network.gamecomm

import com.google.common.flogger.FluentLogger
import l2s.authserver.database.DatabaseFactory
import l2s.commons.dbutils.DbUtils
import l2s.commons.net.HostInfo
import java.sql.Connection
import java.sql.PreparedStatement
import java.util.concurrent.ConcurrentHashMap

/**
 * @reworked by Bonux
 */
class GameServerDescription {

    private val hostMap: MutableMap<Int, HostInfo> = HashMap()
    var serverType = 0
    var ageLimit = 0
    var protocol = 0
    var isOnline = false
    var isPvp = false
    var isShowingBrackets = false
    var isGmOnly = false
    var maxPlayers = 0
    private var _accounts: MutableSet<String> = ConcurrentHashMap.newKeySet()

    constructor() {
    }

    constructor(id: Int, ip: String?, port: Int, key: String?) {
        addHost(HostInfo(id, ip, port, key))
    }

    fun addHost(host: HostInfo) {
        hostMap.put(host.id, host)
    }

    fun removeHost(id: Int): HostInfo? {
        return hostMap.remove(id)
    }

    fun getHost(id: Int): HostInfo? {
        return hostMap.get(id)
    }

    val hosts: Collection<HostInfo>
        get() = hostMap.values

    val online: Int
        get() = _accounts.size

    val accounts: Set<String>
        get() = _accounts

    fun addAccount(account: String) {
        _accounts.add(account)
    }

    fun removeAccount(account: String) {
        _accounts.remove(account)
    }

    fun setDown() {
        isOnline = false
        _accounts.clear()
    }

    fun store(): Boolean {
        var con: Connection? = null
        var statement: PreparedStatement? = null
        try {
            con = DatabaseFactory.getInstance().connection
            for (host in hosts) {
                statement =
                    con.prepareStatement("REPLACE INTO gameservers (`id`, `ip`, `port`, `age_limit`, `pvp`, `max_players`, `type`, `brackets`, `key`) VALUES(?,?,?,?,?,?,?,?,?)")
                var i = 0
                statement.setInt(++i, host.id)
                statement.setString(++i, host.address)
                statement.setShort(++i, host.port.toShort())
                statement.setByte(++i, ageLimit.toByte())
                statement.setByte(++i, (if (isPvp) 1 else 0).toByte())
                statement.setShort(++i, maxPlayers.toShort())
                statement.setInt(++i, serverType)
                statement.setByte(++i, (if (isShowingBrackets) 1 else 0).toByte())
                statement.setString(++i, host.key)
                statement.execute()
                DbUtils.closeQuietly(statement)
            }
        } catch (e: Exception) {
            _log.atSevere().withStackTrace(com.google.common.flogger.StackSize.FULL).withCause(e).log("Error while store gameserver.")
            return false
        } finally {
            DbUtils.closeQuietly(con, statement)
        }
        return true
    }

    companion object {

        private val _log = FluentLogger.forEnclosingClass()

    }

}